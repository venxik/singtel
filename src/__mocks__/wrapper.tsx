import { Provider } from 'react-redux';
import { AppStore, RootState, setupStore } from '../redux';
import { PreloadedState } from '@reduxjs/toolkit';
import { PropsWithChildren } from 'react';
import { render } from '@testing-library/react-native';

interface ExtendedRenderOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function AllTheProviders({ children }: { children: JSX.Element }) {
  return <Provider store={setupStore()}>{children}</Provider>;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
