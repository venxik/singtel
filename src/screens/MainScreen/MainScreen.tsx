import { ScrollView, StyleSheet } from 'react-native';

import { Button, Category, ComicText, View } from '../../components';
import { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import { RootStackNavigationProps } from '../../navigation';
import { useReduxDispatch } from '../../redux';
import { setSelectedCategory } from '../../redux/main';
import { categories } from '../../constants/data';

const MainScreen = () => {
  const [selectCategory, setSelectCategory] = useState('');
  const dispatch = useReduxDispatch();

  const { navigate } = useNavigation<RootStackNavigationProps<'MainScreen'>>();

  const onPressStart = () => {
    dispatch(setSelectedCategory(selectCategory));
    navigate('QuestionScreen');
  };

  return (
    <ScrollView contentContainerStyle={{ flexGrow: 1 }}>
      <View style={styles.container}>
        <ComicText style={styles.title}>Words Puzzle</ComicText>
        <ComicText>Please choose a theme </ComicText>
        <View style={{ height: '50%', justifyContent: 'space-evenly' }}>
          {categories.map((item) => {
            return (
              <Category
                testID={`category-${item}`}
                key={item}
                isSelected={item == selectCategory}
                text={item}
                onPress={() => setSelectCategory(item)}
              />
            );
          })}
        </View>
        <Button
          testID={`button-continue`}
          style={styles.button}
          disabled={selectCategory === ''}
          onPress={onPressStart}>
          <ComicText>Start</ComicText>
        </Button>
        <Button
          testID={`button-leaderboard`}
          style={styles.button}
          onPress={() => navigate('LeaderBoardScreen')}>
          <ComicText>Leaderboards</ComicText>
        </Button>
      </View>
    </ScrollView>
  );
};

export default MainScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
  },
  title: { fontSize: 30, marginVertical: '20%' },
  button: {
    width: '90%',
    marginBottom: 10,
  },
});
