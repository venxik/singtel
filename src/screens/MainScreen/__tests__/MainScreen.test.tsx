import { fireEvent, screen } from '@testing-library/react-native';
import { renderWithProviders } from '../../../__mocks__';
import MainScreen from '../MainScreen';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('MainScreen', () => {
  it('match snapshot', async () => {
    renderWithProviders(<MainScreen />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('can select category button and background color changed', async () => {
    renderWithProviders(<MainScreen />);
    fireEvent.press(screen.getByTestId('category-animals'));
    expect(screen.getByTestId('category-animals')).toHaveStyle({ backgroundColor: '#686fcc' });
  });

  it('button is disabled when category not selected', async () => {
    renderWithProviders(<MainScreen />);
    expect(screen.getByTestId('button-continue')).toHaveAccessibilityState({ disabled: true });
  });

  it('can press start button and navigate to question screen after select category animals', async () => {
    renderWithProviders(<MainScreen />);
    fireEvent.press(screen.getByTestId('category-animals'));
    fireEvent.press(screen.getByTestId('button-continue'));
    expect(mockedNavigate).toHaveBeenCalledWith('QuestionScreen');
  });

  it('can navigate to leaderboard screen', async () => {
    renderWithProviders(<MainScreen />);
    fireEvent.press(screen.getByTestId('button-leaderboard'));
    expect(mockedNavigate).toHaveBeenCalledWith('LeaderboardScreen');
  });
});
