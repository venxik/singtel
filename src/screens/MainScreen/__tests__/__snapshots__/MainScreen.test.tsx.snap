// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen match snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
>
  <View>
    <View
      style={
        [
          {
            "backgroundColor": "#fff",
          },
          {
            "alignItems": "center",
            "flex": 1,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              {
                "fontSize": 30,
                "marginVertical": "20%",
              },
              {
                "fontFamily": "comic",
              },
            ],
          ]
        }
      >
        Words Puzzle
      </Text>
      <Text
        style={
          [
            {
              "color": "#000",
            },
            [
              undefined,
              {
                "fontFamily": "comic",
              },
            ],
          ]
        }
      >
        Please choose a theme 
      </Text>
      <View
        style={
          [
            {
              "backgroundColor": "#fff",
            },
            {
              "height": "50%",
              "justifyContent": "space-evenly",
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 20,
              "borderWidth": 1,
              "opacity": 1,
              "paddingHorizontal": "15%",
              "paddingVertical": "5%",
            }
          }
          testID="category-cities"
        >
          <Text
            style={
              [
                {
                  "color": "#000",
                },
                [
                  {
                    "textAlign": "center",
                  },
                  {
                    "fontFamily": "comic",
                  },
                ],
              ]
            }
            testID="category-text"
          >
            cities
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 20,
              "borderWidth": 1,
              "opacity": 1,
              "paddingHorizontal": "15%",
              "paddingVertical": "5%",
            }
          }
          testID="category-animals"
        >
          <Text
            style={
              [
                {
                  "color": "#000",
                },
                [
                  {
                    "textAlign": "center",
                  },
                  {
                    "fontFamily": "comic",
                  },
                ],
              ]
            }
            testID="category-text"
          >
            animals
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 20,
              "borderWidth": 1,
              "opacity": 1,
              "paddingHorizontal": "15%",
              "paddingVertical": "5%",
            }
          }
          testID="category-foods"
        >
          <Text
            style={
              [
                {
                  "color": "#000",
                },
                [
                  {
                    "textAlign": "center",
                  },
                  {
                    "fontFamily": "comic",
                  },
                ],
              ]
            }
            testID="category-text"
          >
            foods
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ccc",
            "borderRadius": 20,
            "marginBottom": 10,
            "opacity": 1,
            "paddingHorizontal": "15%",
            "paddingVertical": "5%",
            "width": "90%",
          }
        }
        testID="button-continue"
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              [
                undefined,
                {
                  "fontFamily": "comic",
                },
              ],
            ]
          }
        >
          Start
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#686fcc",
            "borderRadius": 20,
            "marginBottom": 10,
            "opacity": 1,
            "paddingHorizontal": "15%",
            "paddingVertical": "5%",
            "width": "90%",
          }
        }
        testID="button-leaderboard"
      >
        <Text
          style={
            [
              {
                "color": "#000",
              },
              [
                undefined,
                {
                  "fontFamily": "comic",
                },
              ],
            ]
          }
        >
          Leaderboards
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
